base58/src/error.rs

```diff
@@ -6,10 +6,53 @@ use core::fmt;
 
 use internals::write_err;
 
+/// An error occurred during base58 decoding (with checksum).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Error(pub(crate) ErrorInner);
+
+impl Error {
+    /// Returns the invalid base58 ssscharacter, if encountered.
+    pub fn invalid_character(&self) -> Option<u8> {
+        match self.0 {
+            ErrorInner::Decode(ref e) => Some(e.invalid_character()),
+            _ => None,
+        }
+    }
+
+    /// Returns the incorrect checksum along with the expected checksum, if encountered.
+    pub fn incorrect_checksum(&self) -> Option<(u32, u32)> {
+        match self.0 {
+            ErrorInner::IncorrectChecksum(ref e) => Some((e.incorrect, e.expected)),
+            _ => None,
+        }
+    }
+
+    /// Returns the invalid base58 string length (require at least 4 bytes for checksum), if encountered.
+    pub fn invalid_length(&self) -> Option<usize> {
+        match self.0 {
+            ErrorInner::TooShort(ref e) => Some(e.length),
+            _ => None,
+        }
+    }
+}
+
+internals::impl_from_infallible!(Error);
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write_err!(f, ""; self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { Some(&self.0) }
```

Comment:
In b4b2bf2a96c630df233d42abd3d22f73f5e878c9:

I don't think we should do this. It's weird to have an extra layer of `source()` that just reveals an inner type, which only exists because of visibility concerns that are only related to this crate.

Reply:
I would suggest dropping all the trait impls on `ErrorInner` and just changing the `Error` impls to use `self.0` instead of `*self`. Smaller diff.

Reply:
Ah sweet, much better idea. Thanks, will do.